software/controller/lib/hal/spi_chain.h

```diff
@@ -0,0 +1,127 @@
+  bool *processed{nullptr};    // boolean we set to inform the caller that the request is processed
+};
+
+template <size_t MaxSlaves>
+class DaisyChain : public Channel, public RxListener {
+ public:
+  DaisyChain(Base spi, DMA::Base dma, const char *name, const char *help_supplement);
+
+  void Initialize(uint8_t null_command, uint8_t reset_command, GPIO::Port clock_port,
+                  uint8_t clock_pin, GPIO::Port miso_port, uint8_t miso_pin, GPIO::Port mosi_port,
+                  uint8_t mosi_pin, GPIO::Port chip_select_port, uint8_t chip_select_pin,
+                  GPIO::Port reset_port, uint8_t reset_pin, uint8_t word_size, Bitrate bitrate);
+
+  bool SendRequest(const Request &request, size_t slave);
+
+  size_t num_slaves() { return num_slaves_; }
+
+  void on_rx_complete() override;
+  void on_rx_error(RxError) override{};
+  void on_character_match() override{};
+
+ protected:
+  // Number of slaves that are actually present in the chain, determined during initilization
+  size_t num_slaves_{0};
+  void ProbeSlaves(uint8_t null_command, uint8_t reset_command);
+
+  // null command we fill the send buffer with when there is nothing to send to a slave
+  uint8_t null_command_{0};
+
+  // Transmission status
+  bool transmitting_data_{false};
+
+  void TransmitNextCommand();
+
+  // We need to handle as many requests queues as we have slaves. We are using normal arrays to do
+  // this as the requests should not stack up, and we are able to empty the queues every so often.
+  static constexpr size_t MaxQueueLength{10};
+
+ private:
+  Request request_queue_[MaxSlaves][MaxQueueLength];
+  size_t queue_count_[MaxSlaves] = {0};
+  size_t current_request_[MaxSlaves] = {0};
+  size_t command_index_[MaxSlaves] = {0};
+  bool end_of_request_[MaxSlaves] = {false};
+  size_t response_count_[MaxSlaves] = {0};
+  bool save_response_[MaxSlaves] = {false};
+
+  // When a command is queued, its data is stored in a local buffer since the scope of the data
+  // being pointed to is not certain.
+  // Length of 20 bytes per slave is arbitrary but should suffice for stepper motors, which take
+  // short commands (length <= 4).
+  // Note that this buffer is shared between slaves
+ protected:
+  static constexpr size_t CommandBufferSize{20 * MaxSlaves};
+  size_t command_buffer_count_{0};
+
+ private:
+  uint8_t command_buffer_[CommandBufferSize] = {0};
+
+  // Buffers used to transmit/receive data from the SPI peripheral
+  uint8_t send_buffer_[MaxSlaves] = {0};
+  uint8_t receive_buffer_[MaxSlaves] = {0};
+
+  void ProcessReceivedData();
+  void SendDataWithBusyWait(uint8_t *command_buffer, uint8_t *response_buffer, size_t length);
+  virtual void WaitResponse();
+
+  Debug::Variable::UInt32 queueing_errors_{"queuing_errors", Debug::Variable::Access::ReadOnly, 0,
+                                           "", "Queueing error counter"};
+};
+
+#include "spi_chain.tpp"
```

Comment:
I would maybe not include it into the namespace but make that namespace explicit in the `tpp` as well?

Reply:
done